keps/sig-api-machinery/3158-API-brotli-compression-support/API-brotli-compression-support.md

```diff
@@ -0,0 +1,105 @@
+# Apiserver support brotli compression
+
+## Table of Contents
+
+<!-- toc -->
+- [Apiserver support brotli compression](#apiserver-support-brotli-compression)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [1.25](#125)
+    - [1.26](#126)
+    - [Implementation Details](#implementation-details)
+      - [How to enable brotli compression](#how-to-enable-brotli-compression)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+Kubernetes sometimes returns extremely large responses to clients outside of its local network, resulting in long delays for components that integrate with the cluster in the list/watch controller pattern. Just as
+https://github.com/kubernetes/enhancements/blob/787e5513d09096756c61aaba1916a73cb1dd348b/keps/sig-api-machinery/2338-graduate-API-gzip-compression-support-to-GA/README.md described.
+
+Brotli has better perfermance than gzip. We should support brotli compression.
+
+## Motivation
+Brotli has better compression ratio and compression time than gzip. Below is my test result:
+
+```shell
+// This is my machine
+Architecture:        x86_64
+CPU(s):              8
+Model name:          Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
+CPU MHz:             2394.454
+```
+Compress 100K kubernetes pods:
+
+| AcceptEncoding | ContentType | Compression Ratio | Compression Time |
+|---|-------------|-------------------|-----------------|
+| gzip    | json        | 35.346268         | 1.308877099s    |
+| brotli    | json        | 42.693703         | 437.817496ms |
+| gzip    | protobuf    | 25.013601         | 842.336524ms    |
+| brotli    | protobuf    | 25.714157         | 433.808134ms |
+
```

Comment:
Just a quick pass (ignoring implementation details for now).

The numbers look great and from that perspective I would really like to see that happening.

That said - I'm not familiar with brotli - so I would like to better understand how the support for that looks like in golang. Pulling that dependency requires the dependency to be very solid one.

I would suggest talking to @liggitt - he might help you with talking about criteria our dependencies need to meet.

Reply:
@wojtek-t @liggitt I use [this package](https://github.com/andybalholm/brotli) which is brotli compressor and decompressor implemented in Go.

Reply:
I would like to get @liggitt or @smarterclayton opinions, but I'm a bit afraid of this dependency because:
- it seem to have pretty low test coverage
- it seems to be maintained by a single person, which is always risky